presto-main/src/main/java/com/facebook/presto/sql/gen/SwitchCodeGenerator.java

```diff
@@ -97,49 +81,84 @@ else if (<value> == <non-constant2>) {
-        elseValue = new BytecodeBlock().visitLabel(nullValue).append(elseValue);
-        // reverse list because current if statement builder doesn't support if/else so we need to build the if statements bottom up
-        for (RowExpression clause : Lists.reverse(whenClauses)) {
+        LabelNode elseLabel = new LabelNode("else");
+        LabelNode endLabel = new LabelNode("end");
+
+        BytecodeBlock block = new BytecodeBlock();
+        BytecodeNode getTempVariableNode = null;
+        if (!searchedCase) {
+            BytecodeNode valueBytecode = generatorContext.generate(value, Optional.empty());
+            Variable tempVariable = scope.createTempVariable(valueType);
+            block.append(valueBytecode)
+                    .append(BytecodeUtils.ifWasNullClearPopAndGoto(scope, elseLabel, void.class, valueType))
+                    .putVariable(tempVariable);
+            getTempVariableNode = VariableInstruction.loadVariable(tempVariable);
+        }
+
+        Variable wasNull = generatorContext.wasNull();
+        block.putVariable(wasNull, false);
+
+        Map<RowExpression, LabelNode> resultLabels = new HashMap<RowExpression, LabelNode>();
+        // We already know the P1 .. Pn are all boolean just call them and search for true (false/null don't matter).
+        for (RowExpression clause : whenClauses) {
             checkArgument(clause instanceof SpecialFormExpression && ((SpecialFormExpression) clause).getForm().equals(WHEN));
 
             RowExpression operand = ((SpecialFormExpression) clause).getArguments().get(0);
+            BytecodeNode operandBytecode;
+
+            if (searchedCase) {
+                operandBytecode = generatorContext.generate(operand, Optional.empty());
+            }
+            else {
+                // call equals(value, operandBytecode)
+                FunctionHandle equalsFunction = generatorContext.getFunctionManager().resolveOperator(EQUAL, fromTypes(value.getType(), operand.getType()));
+                operandBytecode = generatorContext.generateCall(
+                        EQUAL.name(),
+                        generatorContext.getFunctionManager().getBuiltInScalarFunctionImplementation(equalsFunction),
+                        ImmutableList.of(
+                                generatorContext.generate(operand,
+                                Optional.empty()),
+                                getTempVariableNode));
+            }
+
+            block.append(operandBytecode);
+
+            IfStatement ifWasNull = new IfStatement();
+            ifWasNull.condition(wasNull)
+                    .ifTrue()
+                    .putVariable(wasNull, false)
+                    .pop(Boolean.class); // pop the result of the predicate eval
+
+            // Here the TOS  is the result of the predicate.
             RowExpression result = ((SpecialFormExpression) clause).getArguments().get(1);
+            LabelNode target = resultLabels.get(result);
```

Comment:
`resultLabels.putIfAbsent(...)`?

Reply:
As you can see I left Java at jdk 5 :) thanks for teaching me some new/cool patterns!

Reply:
Except in this case it's a bit awkward to do it because of the new call (and also putIfAbsent returning null). So I'm just keeping the onld one.